File: web/types/adk-schema.ts
/**
 * TypeScript type definitions for SRE Agent data structures.
 * These types mirror the Pydantic schemas from the backend.
 */

// =============================================================================
// Enums
// =============================================================================

export type Severity = "critical" | "high" | "medium" | "low" | "info";
export type Confidence = "high" | "medium" | "low";
export type ToolStatus = "success" | "error" | "partial";
export type OverallAssessment = "healthy" | "degraded" | "critical";
export type ChangeType = "added" | "removed" | "modified" | "depth_change" | "fanout_change";
export type EffectType = "root_cause" | "direct_effect" | "cascaded_effect";
export type ImpactType = "latency" | "error_rate" | "throughput" | "availability";

// =============================================================================
// Base Types
// =============================================================================

export interface BaseToolResponse<T = unknown> {
  status: ToolStatus;
  result?: T;
  error?: string;
  metadata: Record<string, unknown>;
}

// =============================================================================
// Trace Types
// =============================================================================

export interface SpanInfo {
  span_id: string;
  name: string;
  duration_ms?: number;
  parent_span_id?: string | null;
  has_error: boolean;
  labels: Record<string, string>;
  start_time?: string;
  end_time?: string;
  service_name?: string;
  status_code?: number | string;
}

export interface Trace {
  trace_id: string;
  project_id?: string;
  spans: SpanInfo[];
  start_time: string;
  end_time: string;
  total_duration_ms: number;
  root_span_id?: string;
}

export interface TraceSummary {
  trace_id: string;
  span_count: number;
  total_duration_ms: number;
  error_count: number;
  max_depth: number;
}

export interface LatencyDiff {
  span_name: string;
  baseline_ms: number;
  target_ms: number;
  diff_ms: number;
  diff_percent: number;
  severity: Severity;
}

export interface ErrorDiff {
  span_name: string;
  error_type: string;
  error_message?: string;
  status_code?: number | string;
  is_new: boolean;
}

export interface StructureDiff {
  change_type: ChangeType;
  span_name: string;
  description: string;
}

export interface TraceComparisonReport {
  baseline_summary: TraceSummary;
  target_summary: TraceSummary;
  overall_assessment: OverallAssessment;
  latency_findings: LatencyDiff[];
  error_findings: ErrorDiff[];
  structure_findings: StructureDiff[];
  root_cause_hypothesis: string;
  recommendations: string[];
}

// =============================================================================
// Log Types
// =============================================================================

export interface LogEntry {
  timestamp: string;
  severity: string;
  payload: string;
  resource: Record<string, unknown>;
  labels?: Record<string, string>;
}

export interface LogPattern {
  pattern_id: string;
  template: string;
  count: number;
  first_seen?: string;
  last_seen?: string;
  severity_counts: Record<string, number>;
  sample_messages: string[];
  resources: string[];
}

export interface LogPatternSummary {
  total_logs_processed: number;
  unique_patterns: number;
  severity_distribution: Record<string, number>;
  compression_ratio: number;
  top_patterns: LogPattern[];
  error_patterns: LogPattern[];
}

export interface PatternComparison {
  new_patterns: LogPattern[];
  disappeared_patterns: LogPattern[];
  increased_patterns: Array<{ pattern: LogPattern; increase_pct: number }>;
  decreased_patterns: Array<{ pattern: LogPattern; decrease_pct: number }>;
  stable_patterns_count: number;
}

export interface LogComparisonResult {
  baseline_summary: {
    total_logs: number;
    unique_patterns: number;
  };
  comparison_summary: {
    total_logs: number;
    unique_patterns: number;
  };
  anomalies: PatternComparison;
  alert_level: string;
}

// =============================================================================
// Metrics Types
// =============================================================================

export interface TimeSeriesPoint {
  timestamp: string;
  value: number;
}

export interface TimeSeries {
  metric: Record<string, unknown>;
  resource: Record<string, unknown>;
  points: TimeSeriesPoint[];
  metric_name?: string;
}

export interface Anomaly {
  timestamp: string;
  value: number;
  expected_value?: number;
  deviation?: number;
  severity: Severity;
  description?: string;
}

export interface MetricWithAnomalies {
  series: TimeSeries;
  anomalies: Anomaly[];
  incident_window?: {
    start: string;
    end: string;
  };
}

export interface LatencyDistribution {
  sample_size: number;
  mean_ms: number;
  median_ms: number;
  p90_ms: number;
  p95_ms: number;
  p99_ms: number;
  std_dev_ms: number;
  coefficient_of_variation: number;
}

// =============================================================================
// Analysis Types
// =============================================================================

export interface AnomalyFinding {
  span_name: string;
  observed_ms: number;
  expected_ms: number;
  z_score: number;
  severity: Severity;
}

export interface CriticalPathSegment {
  span_name: string;
  duration_ms: number;
  percentage_of_total: number;
  is_optimization_target: boolean;
}

export interface StatisticalAnalysisReport {
  latency_distribution: LatencyDistribution;
  anomaly_threshold: number;
  anomalies: AnomalyFinding[];
  critical_path: CriticalPathSegment[];
  optimization_opportunities: string[];
}

export interface RootCauseCandidate {
  rank: number;
  span_name: string;
  slowdown_ms: number;
  confidence: Confidence;
  reasoning: string;
}

export interface CausalChainLink {
  span_name: string;
  effect_type: EffectType;
  latency_contribution_ms: number;
}

export interface CausalAnalysisReport {
  causal_chain: CausalChainLink[];
  root_cause_candidates: RootCauseCandidate[];
  propagation_depth: number;
  primary_root_cause: string;
  confidence: Confidence;
  conclusion: string;
  recommended_actions: string[];
}

export interface ServiceImpact {
  service_name: string;
  impact_type: ImpactType;
  severity: Severity;
  baseline_value: number;
  current_value: number;
  change_percent: number;
  affected_operations: string[];
}

export interface ServiceImpactReport {
  total_services_analyzed: number;
  impacted_services_count: number;
  service_impacts: ServiceImpact[];
  cross_service_effects: string[];
  blast_radius_assessment: string;
}

// =============================================================================
// Remediation Types
// =============================================================================

export interface RemediationStep {
  action: string;
  description: string;
  steps: string[];
  risk: "low" | "medium" | "high";
  effort: "low" | "medium" | "high";
  category?: string;
  source_pattern?: string;
}

export interface RemediationSuggestion {
  matched_patterns: string[];
  categories?: string[];
  finding_summary?: string;
  suggestions: RemediationStep[];
  recommended_first_action?: RemediationStep;
  quick_wins: RemediationStep[];
}

export interface GcloudCommand {
  description: string;
  command: string;
}

export interface RemediationCommands {
  remediation_type: string;
  resource: string;
  project: string;
  commands: GcloudCommand[];
  warning: string;
}

export interface RiskAssessment {
  action: string;
  service: string;
  change: string;
  risk_assessment: {
    level: "low" | "medium" | "high";
    confidence: string;
    factors: string[];
  };
  recommendations: {
    proceed: boolean;
    require_approval: boolean;
    mitigations: string[];
  };
  checklist: string[];
}

// =============================================================================
// UI State Types
// =============================================================================

export type AgentType =
  | "orchestrator"
  | "latency_specialist"
  | "error_analyst"
  | "log_pattern_engine"
  | "metrics_correlator"
  | "remediation_advisor"
  | "idle";

export interface AgentStatus {
  currentAgent: AgentType;
  message: string;
  progress?: number;
  startTime?: string;
}

export type CanvasView =
  | { type: "empty" }
  | { type: "trace"; data: Trace }
  | { type: "trace_comparison"; data: TraceComparisonReport }
  | { type: "log_patterns"; data: LogPatternSummary }
  | { type: "log_comparison"; data: LogComparisonResult }
  | { type: "metrics"; data: MetricWithAnomalies }
  | { type: "remediation"; data: RemediationSuggestion }
  | { type: "risk_assessment"; data: RiskAssessment }
  | { type: "service_impact"; data: ServiceImpactReport }
  | { type: "causal_analysis"; data: CausalAnalysisReport };

// =============================================================================
// Error Types
// =============================================================================

export interface ErrorInfo {
  span_name: string;
  error_type: string;
  status_code?: number | string;
  error_message?: string;
  service_name?: string;
}

export interface ErrorAnalysisReport {
  baseline_error_count: number;
  target_error_count: number;
  net_change: number;
  new_errors: ErrorInfo[];
  resolved_errors: ErrorInfo[];
  common_errors: ErrorInfo[];
  error_pattern_analysis: string;
  recommendations: string[];
}

File: autosre/lib/models/adk_schema.dart

class SpanInfo {
  final String spanId;
  final String traceId;
  final String name;
  final DateTime startTime;
  final DateTime endTime;
  final Map<String, dynamic> attributes;
  final String status; // 'OK', 'ERROR'
  final String? parentSpanId;

  SpanInfo({
    required this.spanId,
    required this.traceId,
    required this.name,
    required this.startTime,
    required this.endTime,
    required this.attributes,
    required this.status,
    this.parentSpanId,
  });

  Duration get duration => endTime.difference(startTime);

  factory SpanInfo.fromJson(Map<String, dynamic> json) {
    return SpanInfo(
      spanId: json['span_id'],
      traceId: json['trace_id'],
      name: json['name'],
      startTime: DateTime.parse(json['start_time']),
      endTime: DateTime.parse(json['end_time']),
      attributes: Map<String, dynamic>.from(json['attributes'] ?? {}),
      status: json['status'] ?? 'OK',
      parentSpanId: json['parent_span_id'],
    );
  }
}

class Trace {
  final String traceId;
  final List<SpanInfo> spans;

  Trace({required this.traceId, required this.spans});

  factory Trace.fromJson(Map<String, dynamic> json) {
    var list = json['spans'] as List;
    List<SpanInfo> spansList = list.map((i) => SpanInfo.fromJson(i)).toList();
    return Trace(traceId: json['trace_id'], spans: spansList);
  }
}

class MetricPoint {
  final DateTime timestamp;
  final double value;
  final bool isAnomaly;

  MetricPoint({required this.timestamp, required this.value, this.isAnomaly = false});

  factory MetricPoint.fromJson(Map<String, dynamic> json) {
    return MetricPoint(
      timestamp: DateTime.parse(json['timestamp']),
      value: (json['value'] as num).toDouble(),
      isAnomaly: json['is_anomaly'] ?? false,
    );
  }
}

class MetricSeries {
    final String metricName;
    final List<MetricPoint> points;
    final Map<String, dynamic> labels;

    MetricSeries({required this.metricName, required this.points, required this.labels});

    factory MetricSeries.fromJson(Map<String, dynamic> json) {
         var list = json['points'] as List;
         List<MetricPoint> pointsList = list.map((i) => MetricPoint.fromJson(i)).toList();
         return MetricSeries(
             metricName: json['metric_name'],
             points: pointsList,
             labels: Map<String, dynamic>.from(json['labels'] ?? {}),
         );
    }
}

class LogPattern {
    final String template;
    final int count;
    final Map<String, int> severityCounts;

    LogPattern({required this.template, required this.count, required this.severityCounts});

    factory LogPattern.fromJson(Map<String, dynamic> json) {
        return LogPattern(
            template: json['template'],
            count: json['count'],
            severityCounts: Map<String, int>.from(json['severity_counts'] ?? {}),
        );
    }
}

class RemediationStep {
    final String command;
    final String description;

    RemediationStep({required this.command, required this.description});

    factory RemediationStep.fromJson(Map<String, dynamic> json) {
        return RemediationStep(
            command: json['command'],
            description: json['description'],
        );
    }
}

class RemediationPlan {
    final String issue;
    final String risk; // 'low', 'medium', 'high'
    final List<RemediationStep> steps;

    RemediationPlan({required this.issue, required this.risk, required this.steps});

    factory RemediationPlan.fromJson(Map<String, dynamic> json) {
        var list = json['steps'] as List;
        List<RemediationStep> stepsList = list.map((i) => RemediationStep.fromJson(i)).toList();
        return RemediationPlan(
            issue: json['issue'],
            risk: json['risk'],
            steps: stepsList,
        );
    }
}

class ToolLog {
    final String toolName;
    final Map<String, dynamic> args;
    final String status; // 'running', 'completed', 'error'
    final String? result;
    final String? timestamp;

    ToolLog({
        required this.toolName,
        required this.args,
        required this.status,
        this.result,
        this.timestamp,
    });

    factory ToolLog.fromJson(Map<String, dynamic> json) {
        return ToolLog(
            toolName: json['tool_name'],
            args: Map<String, dynamic>.from(json['args'] ?? {}),
            status: json['status'] ?? 'unknown',
            result: json['result']?.toString(), // Handle both string and complex object results by stringifying for now
            timestamp: json['timestamp'],
        );
    }
}

/// Individual log entry with full payload for expandable JSON view
class LogEntry {
    final String insertId;
    final DateTime timestamp;
    final String severity; // 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'
    final dynamic payload; // Can be String (text) or Map (JSON)
    final Map<String, String> resourceLabels;
    final String resourceType;
    final String? traceId;
    final String? spanId;
    final Map<String, dynamic>? httpRequest;

    LogEntry({
        required this.insertId,
        required this.timestamp,
        required this.severity,
        required this.payload,
        required this.resourceLabels,
        required this.resourceType,
        this.traceId,
        this.spanId,
        this.httpRequest,
    });

    bool get isJsonPayload => payload is Map;

    String get payloadPreview {
        if (payload is String) {
            return payload.length > 200 ? '${payload.substring(0, 200)}...' : payload;
        }
        if (payload is Map) {
            final message = payload['message'] ?? payload['msg'] ?? payload['text'];
            if (message != null) return message.toString();
            return payload.toString().length > 200
                ? '${payload.toString().substring(0, 200)}...'
                : payload.toString();
        }
        return payload?.toString() ?? '';
    }

    factory LogEntry.fromJson(Map<String, dynamic> json) {
        return LogEntry(
            insertId: json['insert_id'] ?? '',
            timestamp: DateTime.parse(json['timestamp']),
            severity: json['severity'] ?? 'INFO',
            payload: json['payload'],
            resourceLabels: Map<String, String>.from(json['resource_labels'] ?? {}),
            resourceType: json['resource_type'] ?? 'unknown',
            traceId: json['trace_id'],
            spanId: json['span_id'],
            httpRequest: json['http_request'] != null
                ? Map<String, dynamic>.from(json['http_request'])
                : null,
        );
    }
}

/// Container for log entries viewer
class LogEntriesData {
    final List<LogEntry> entries;
    final String? filter;
    final String? projectId;
    final String? nextPageToken;

    LogEntriesData({
        required this.entries,
        this.filter,
        this.projectId,
        this.nextPageToken,
    });

    factory LogEntriesData.fromJson(Map<String, dynamic> json) {
        final entriesList = (json['entries'] as List? ?? [])
            .map((e) => LogEntry.fromJson(Map<String, dynamic>.from(e)))
            .toList();
        return LogEntriesData(
            entries: entriesList,
            filter: json['filter'],
            projectId: json['project_id'],
            nextPageToken: json['next_page_token'],
        );
    }
}
